nupic/research/frameworks/vernon/mixins/dendrite_metrics.py

```diff
@@ -0,0 +1,247 @@
+from nupic.research.frameworks.pytorch.model_utils import filter_modules
+
+__all__ = [
+    "PlotDendriteMetrics",
+]
+
+
+class PlotDendriteMetrics(metaclass=abc.ABCMeta):
+    """
+    This is a mixin for creating custom plots of metrics for
+    apply-dendrite modules (those of type `ApplyDendritesBase`_). The user defines and
+    gives a plotting function which is then called on the following arguments
+
+        - dendrite_activations: the input activations passed to the apply-dendrites
+                                module; these are meant to be the output of a
+                                `DendriteSegments` module; they will be of shape
+                                batch_size x num_units x num_segments
+        - winning_mask: the mask of the winning segments, those chosen by the
+                        apply-dendrites modules; this will be of shape
+                        batch_size x num_units
+        - targets: the targets that correspond to each sample in the batch
+
+    Plots can be configured to use fewer samples (helpful for plotting a small batches
+    of individual samples) and to plot every so many epochs (so that training isn't
+    slowed down too much). Whenever a plot is made, the raw data used to create it s
+    saved so it may be reproduced and edited off-line.
+
+    .. warning:: When using this mixin with Ray, be careful to have 'plot_func' return
+                 an object that can be logged. Often, Ray will attempt to create a
+                 deepcopy prior to logging which can't be done on most plots. Try
+                 wrapping the plot as done in `prep_plot_for_wandb`_.
+
+    .. _ApplyDendritesBase: nupic/reasearch/frameworks/dendrites/modules
+    .. _prep_plot_for_wandb: nupic/reasearch/frameworks/wandb/ray_wandb
+
+
+    :param config:
+        - plot_dendrite_metrics_args:
+
+            # The following arguments specify which modules of type `ApplyDendritesBase`
+            # to track and create plots. The same modules are used for each 'plot_func'.
+            - include_modules: (optional) a list of module types to track
+            - include_names: (optional) a list of module names to track e.g.
+                             "features.stem"
+            - include_patterns: (optional) a list of regex patterns to compare to the
+                                names; for instance, all feature parameters in ResNet
+                                can be included through "features.*"
+
+            <plot_name>: This can be any string and maps to a dictionary of arguments
+                         for each desired plot. The resulting plot will be logged under
+                         "<plot_name>/<module_name>" in the results dictionary.
+
+                - plot_func: the function called for plotting; must take three
+                             arguments: 'dendrite_activations', 'winning_mask', and
+                             'targets' (see above)
```

Comment:
Recall that the plotting functions [here](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/dendrites/metrics/plotting.py) don't all need all 3 arguments. E.g., the function `plot_percent_active_dendrites` doesn't take `dendrite_activations`, so this seems inconsistent with what the docstring is saying.

Reply:
It supports the most general case which takes three arguments. My thinking was that we can adjust the signature of the functions as needed. (see below)

Reply:
The mixin supports the more general case that takes all three. I figure we can adjust the signatures as needed (see below).